source/slang/slang-ir-pytorch-cpp-binding.cpp

```diff
@@ -858,6 +859,99 @@ IRFunc* generateCUDAWrapperForFunc(IRFunc* func, DiagnosticSink* sink)
     return hostFunc;
 }
 
+void lowerBuiltinTypesForKernelEntryPoints(IRModule* module, DiagnosticSink*)
+{
+    List<IRFunc*> cudaKernels;
+    for (auto globalInst : module->getGlobalInsts())
+    {
+        if (auto func = as<IRFunc>(globalInst))
+        {
+            if (func->findDecoration<IRCudaKernelDecoration>())
+            {
+                cudaKernels.add(func);
+            }
+        }
+    }
+
+    Dictionary<IRType*, LoweredBuiltinTypeInfo> typeMap;
+    IRBuilder builder(module);
+    for (auto func : cudaKernels)
+    {
+        // Go through parameters and replace any built-in types with their equivalent.
+        List<IRParam*> params;
+        for (auto param : func->getFirstBlock()->getParams())
+        {
+            params.add(param);
+        }
+
+        bool changed = false;
+        for (auto param : params)
+        {
+            LoweredBuiltinTypeInfo info = LoweredBuiltinTypeInfo();
```

Comment:
This is not handling struct types?

Can we extend the existing lower-buffer-element-type pass to make it work for CUDA?

Reply:
I tried to adapt the lower-buffer-element pass to also process this, but there were too many thing that had to be done differentl,y. For one, the buffer lowering pass needs to care about matrix and array types, and expresses them in terms of vectors. For builtin-type lowering we need to lower matrix and vector types into structs of arrays, and we don't necessarily care about alignment (since this pass is `[CUDAKernel]` specific). 
We also don't need to care about rewriting accesses, since we can construct the built-in type back on the CUDA side & retain all the other instructions.

But yes, we do need to handle built-in types that are nested inside structs.

Reply:
That's what I mean by extending the pass. We need to implement the logic to selectively do certain things based on the target we are generating code for, but share the overall structure/flow of the pass so we don't have to duplicate the logic for e.g. handling arrays and structures.

Reply:
I can give it another try, but even the structure of the rewriting pass needs to be a bit different. 

The current rewriting pass finds buffer types and works backwards from there (affecting all uses of the type). We need a separate, more narrowly scoped pathway which only affects the parameter type of a kernel function (and not other uses of the type).  

Technically, we can group the two passes together, but I felt that since both the traversal logic and rewrite logic was different, it might be more confusing to have them in the same step. 
My intention was to instead group this rewrite pass with the other type rewriting passes in the auto-binding logic (we already have a device-to-host type translation step, and a host-to-tuple-type translation step). 

Would that work better, or should I try again to unify this logic with the buffer lowering pass?

Reply:
I see. I'll trust your judgement on how much overlap there is to worth the unification. I think what we need is a pass that can rewrite the type of a global variable, a struct field, or a function parameter. This pass can then be used to drive both rewriting we are dealing with here. For AutoPyBind, you only enable the worklist discovery logic for entry point parameters, and not anything else.

Reply:
It is true that we need to unify type rewriting steps though.. I feel like I've written 3 or 4 different passes that almost work the same way, but not quite. 
I'll raise a separate issue and try to merge these passes together (starting with the auto-bind related rewrites) as a part of the cleanup later this week.